src/drivers/gitlab.js

```diff
@@ -230,6 +230,30 @@ class Gitlab {
     });
   }
 
+  async pipelineRestart(opts = {}) {
+    const projectPath = await this.projectPath();
+    const { jobId } = opts;
+
+    const {
+      pipeline: { id }
+    } = await this.request({
+      endpoint: `/projects/${projectPath}/jobs/${jobId}`
+    });
+
+    let status;
+    while (!status || status === 'running')
+      ({ status } = await this.request({
+        endpoint: `/projects/${projectPath}/pipelines/${id}/cancel`,
+        method: 'POST'
+      }));
+
+    while (status !== 'running')
+      ({ status } = await this.request({
+        endpoint: `/projects/${projectPath}/pipelines/${id}/retry`,
+        method: 'POST'
+      }));
```

Comment:
Is there any way of making sure that the job has been cancelled or retried that doesn't involve issuing a burst of API calls repeating the original request? Asking once for the pipeline to stop/retry should be enough, and then it would be matter of using the more polite [`GET /projects/:id/pipelines/:pipeline_id`](https://docs.gitlab.com/ee/api/pipelines.html#get-a-single-pipeline) to query the status or throw an error if it hasn't changed after a reasonable timeout.

Reply:
I love how this function is written, ❤️  but the API requests it performs don't seem especially polite towards the GitLab backend.

Reply:
> [...] _the request may take half a second? so we are far below their limits._

Ideally, rate limiting delays should be explicit and not account for request times. What if GitLab uses yet unknown superpowers to speed up their API so it takes a fraction of the time it used to take for this specific request? Again, those are words from the "deffensive programmer" I bear deep inside me. 🤫 

> _I prefer to ensure that we can trigger the workflow before die in the 30 secs limit that we have_

- Is this endpoint supposed to be unreliable?
- What's the difference between calling it a single time or a hundred times? Is it idempotent?

Reply:
> Is it idempotent?

Its not or apparently not. REsponse comes with current state and after a cancel the response might come with a running, still.

Reply:
Additionally the post comes with the state as GET so I can trigger the cancel to verify also the state instead of a GET

Reply:
> _Response comes with current state and after a cancel the response might come with a running, still._

Yes, that's what I guessed after reading your code. I was wondering if issuing the `POST` request repeatedly would produce a different behavior. As far as I know, issuing the request twice won't expedite the action and may even cause issues with some backends (not saying that it's the case of GitLab though).

Also, there wouldn't make any sense to send the request multiple times as a failure prevention measure, because the loop would `throw` and break after the first error.

> _Additionally the post comes with the state as GET so I can trigger the cancel to verify also the state instead of a GET_

Yes, it's elegant and convenient from the code standpoint, but my concern is that the behavior towards the API might be suboptimal.

Reply:
Its going to be the same @0x2b3bfa0 doing POST than GET and at least posting I ensure it.

Reply:
@DavidGOrtega, under your sole responsibility. 🤷🏼‍♂️  If the endpoint is idempotent, I guess that it's just an inconsequential hack and won't make any difference to the server.

Reply:
I think you don't fully understand me  @0x2b3bfa0. I do the post also to retrieve the status. It's the basically the same. As I said the call is NOT idempotent. Or I cant ensure that the stoppage will happen, the ONLY thing that ensures the stoppage is having it in the response.

Reply:
> _I cant ensure that the stoppage will happen_

Do you mean that a single call to `POST /projects/:id/pipelines/:pipeline_id/cancel` isn't going to cancel the pipeline anytime in the future? It appear like, even if the action isn't immediate, it gets scheduled and applied in matter of seconds.

> _the ONLY thing that ensures the stoppage is having it in the response_

If you mean that the only way of making sure that it was effectively cancelled would be repeadtedly checking for the status until it matches the desired one, yes, I understand what you're saying. On the other hand, if you mean that the only way of querying the status is repeating the `POST` request, I would say that that's not the case: it might be elegant code-wise, but a bit careless API-wise.

Reply:
To be honest and as I said before it/s careful with the API anyway. Im polling until having the response I'm expecting and Im not going to "guess" it was ok or not. While Im under Gitlab's  terms of good usage its simply perfect.
If we were abusing the API I would have slow down the request but the solution would have been EXACTLY the same

Reply:
> _the solution would have been EXACTLY the same_

I would have used [`GET /projects/:id/pipelines/:pipeline_id`](https://docs.gitlab.com/ee/api/pipelines.html#get-a-single-pipeline) to query the status instead of repeating the `POST /projects/:id/pipelines/:pipeline_id/cancel` command each time, that's my point.

Anyhow, https://github.com/iterative/cml/pull/583#discussion_r659956794. As long as it works and the backend undestands your intent, everything should be fine.

Reply:
As I said the GET is the same as POST for our purposes. Same polling. Same politeness. The main difference is calling the cancel. If we had multiple clients checking the same endpoint they would not be able to determine that the state is terminating... or terminated or whatever. 
They would have to do TWO calls at least GET and POST if the response is 'running' even that another client might have raise it.
What ever...

Reply:
> everything should be fine.

Its indeed the most effective code for our purposes in term of amount of code and ensure the correctness